### PR TITLE
New rule S2053: Hashes should include an unpredictable salt - for VB.NET

### LINKED ISSUE #7560: New rule S2053: Hashes should include an unpredictable salt - for VB.NET
Once [HashesShouldHaveUnpredictableSalt is migrated to the new SE engine](https://github.com/SonarSource/sonar-dotnet/issues/7558), implement the VB.NET version of the rule:
  * Create VB.NET test cases (based on the C# test cases)
  * Set `ShouldExecute` to always return true
  * Modify rule logic if necessary to accommodate VB.NET test cases
  * Update ITs
  * Implement `ShouldExecute` for VB.NET

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #7560

### CODE REVIEW

**Comment by zsolt-kolbay-sonarsource** [comment]
The RSPEC has been updated by AppSec, but the title wasn't changed by mistake.
The specs need to be clarified before this can be merged. Until then, the PR remains a Draft.

File: analyzers/src/SonarAnalyzer.VisualBasic/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            if (!Result && node is ObjectCreationExpressionSyntax)
+            {
+                Result = true;
+            }
+            else
+            {
+                base.Visit(node);
+            }
+        }
+    }
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.cs:51]: Similar comment here - there's no need to check in the Visit. There's a dedicated visitor for object creation.

```vbnet
        public override void Visit(SyntaxNode node)
        {
            if (!Result)
            {
                base.Visit(node);
            }
        }

        public override void VisitObjectCreationExpression(ObjectCreationExpressionSyntax node) =>
            Result = true;
```

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
Looks ok - but it's good to revisit mainly the tests after we decide what to do with the new info from appsec.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.vb [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+Class Program
+    Private Const passwordString As String = "Secret"
+    Private cspParams As CspParameters = New CspParameters()
+    Private ReadOnly passwordBytes As Byte() = Encoding.UTF8.GetBytes(passwordString)
+
+    ' Out of the two issues (salt is too short vs. salt is predictable) being predictable is the more serious one.
+    ' If both issues are present then the rule's message will reflect on the salt being predictable.
+    Public Sub ShortAndConstantSaltIsNotCompliant()
+        Dim shortAndConstantSalt = New Byte(14) {}
+        Dim pdb1 = New PasswordDeriveBytes(passwordBytes, shortAndConstantSalt)                                                      ' Noncompliant {{Make this salt unpredictable.}}
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.vb:15]: Could you please add an object creation with all capitals - as VB.NEt is case insensitive?
There should be no issue but better be sure.

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM - just add an extra text for completion.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.vb (hunk 2 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/HashesShouldHaveUnpredictableSalt.vb:15]: Good catch! It got removed by auto-formatting 😃

**Pull request merged**